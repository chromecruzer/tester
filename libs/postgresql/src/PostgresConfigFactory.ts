import * as fs from "fs-extra";

export interface PostgresConfig {
  user: string;
  password: string;
  port: number;
  host: string;
  rootdatabase: string;
  database: string;
  schema: string;
  audit_schema: string;
}

export default class PostgresConfigFactory {
  constructor(public config: PostgresConfig) {
  }

  public static async load(filepath, nodeFs = fs) {
    // console.log(`opening dbConfig at ${filepath}`);
    const config = await nodeFs.readJSON(filepath);
    return new PostgresConfigFactory(config);
  }

  public getTracConfig() {
    const {user, password, port, host, database} = this.config;
    return {
      user,
      password,
      port,
      host,
      database
    }
  }

  public getRootConfig() {
    const {user, password, port, host, rootdatabase} = this.config;
    return {
      user,
      password,
      port,
      host,
      database: rootdatabase
    }
  }

  public getSchemaName() {
    return this.config.schema;
  }

  public getAuditSchemaName() {
    return this.config.audit_schema;
  }

  public getDbName() {
    return this.config.database;
  }
}
