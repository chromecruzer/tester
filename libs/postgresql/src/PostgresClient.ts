import {Pool} from "pg";
import PostgresConfigFactory from "./PostgresConfigFactory";

export default class PostgresClient {
  private pool: Pool;
  constructor(public config: PostgresConfigFactory) {
    this.pool = new Pool(this.config.getTracConfig());
  }

  public getClient() {
    return this.pool.connect();
  }

  public release(client) {
    console.log('disconnecting client');
    client.release(true);

  }
  public disconnect() {
    return this.pool.end();
  }

  public async throwError(stage, err, client) {
    this.release(client);
    return Promise.reject(Error(`${stage} produced an error ${err}`));
  }
}
