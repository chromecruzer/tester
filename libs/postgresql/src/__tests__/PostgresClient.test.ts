import PostgresConfigFactory from "../PostgresConfigFactory";
import PostgresClient from "../PostgresClient";

describe('PostgresClient', () => {
  let mockPool, mockClient, mockConfig, tracClient;
  beforeEach(()=> {
    mockPool = {
      connect: jest.fn(),
      end: jest.fn(),
    };
    mockClient = {
      release: jest.fn()
    }
    mockPool.connect.mockImplementation(() => {
      return Promise.resolve(mockClient);
    })
    mockPool.end.mockImplementation(() => {
      return Promise.resolve('end worked');
    })
    mockConfig = new PostgresConfigFactory({
      user: 'db user',
      password: 'user password',
      port: 4000,
      rootdatabase: 'db root',
      database: 'user db',
      host: 'hostname',
      audit_schema: 'audit schema',
      schema: 'db schema'
    });
    tracClient = new PostgresClient(mockConfig);
    tracClient.pool = mockPool;
  });
  describe('#getClient', () => {
    it('should return a promise with success', () => {
      return tracClient.getClient().then(result => {
        // console.log('checking results');
        expect(result).toEqual(mockClient);
      })
    });
  });
  describe('#release', () => {
    it('should return a promise with success', () => {
      return tracClient.release(mockClient)
    });

  });
  describe('throwError', () => {
    it('should throw error with specified contents', () => {
      return expect(tracClient.throwError('stage','I quit because that error bugs me',
        mockClient)).rejects.toMatchSnapshot();
    });
  });
});
