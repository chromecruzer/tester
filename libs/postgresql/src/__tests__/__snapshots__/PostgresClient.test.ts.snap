// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostgresClient throwError should throw error with specified contents 1`] = `[Error: stage produced an error I quit because that error bugs me]`;
